src/RwaJob.sol

```diff
@@ -0,0 +1,149 @@
+    function list() external view returns (bytes32[] memory);
+    function getComponent(bytes32 ilk, bytes32 name) external view returns (address addr, uint88 variant);
+}
+
+interface RwaJarLike {
+    function void() external;
+}
+
+interface RwaUrnLike {
+    function wipe(uint256 wad) external;
+}
+
+/**
+ * @title Trigger void of RWA Jars that have a minimum balance threshold
+ * @author David Krett <
+ */
+contract RwaJob is IJob {
+    SequencerLike public immutable sequencer;
+    RWARegistryLike public immutable rwaRegistry;
+    DaiAbstract public immutable dai;
+    uint256 public immutable threshold;
+
+    bytes32 constant internal URN_COMPONENT = "urn";
+    bytes32 constant internal JAR_COMPONENT = "jar";
+
+    // --- Errors ---
+    error NotMaster(bytes32 network);
+    error DealNotActive(bytes32 ilk);
+    error BalanceBelowThreshold(bytes32 ilk, bytes32 componentName, uint256 balance);
+    error InvalidComponent(bytes32 ilk, bytes32 componentName);
+
+    // --- Events ---
+    event Work(bytes32 indexed network, bytes32 indexed ilk, bytes32 indexed componentName);
+
+    constructor(address _sequencer, address _rwaRegistry, address _dai, uint256 _threshold) {
+        sequencer = SequencerLike(_sequencer);
+        rwaRegistry = RWARegistryLike(_rwaRegistry);
+        threshold = _threshold;
+        dai = DaiAbstract(_dai);
+    }
+
+    function work(bytes32 network, bytes calldata args) external override {
+        if (!sequencer.isMaster(network)) revert NotMaster(network);
+
+        (bytes32 ilk, bytes32 componentName) = abi.decode(args, (bytes32, bytes32));
+
+        (RWARegistryLike.DealStatus status,) = rwaRegistry.ilkToDeal(ilk);
+
+        // If the deal is not active, skip it.
+        if (status != RWARegistryLike.DealStatus.ACTIVE) {
+            revert DealNotActive(ilk);
+        }
+
+        (address addr, uint88 variant) = rwaRegistry.getComponent(ilk, componentName);
+
+        if (componentName == JAR_COMPONENT) {
+            executeJarActions(ilk, addr, variant);
+        } else if(componentName == URN_COMPONENT) {
+            executeUrnActions(ilk, addr, variant);
+        } else {
+            revert InvalidComponent(ilk, componentName);
+        }
+
+        emit Work(network, ilk, componentName);
+    }
+
+    function executeJarActions(bytes32 ilk, address jar, uint88) internal {
+        // If its balance is below the threshold, skip it.
+        if (dai.balanceOf(jar) < threshold) {
+            revert BalanceBelowThreshold(ilk, JAR_COMPONENT, dai.balanceOf(jar));
+        }
+
+        RwaJarLike(jar).void();
+    }
+
+    function executeUrnActions(bytes32 ilk, address urn, uint88) internal {
+        // If its balance is below the threshold, skip it.
+        if (dai.balanceOf(urn) < threshold) {
+            revert BalanceBelowThreshold(ilk, URN_COMPONENT, dai.balanceOf(urn));
+        }
+
+        RwaUrnLike(urn).wipe(dai.balanceOf(urn));
+    }
```

Comment:
This logic is repeated. You can just do in the `work` function:

```
// If its balance is below the threshold, skip it.
uint256 bal = dai.balanceOf(addr);
if (bal < threshold) {
    revert BalanceBelowThreshold(ilk, componentName, bal);
}

if (componentName == JAR_COMPONENT) {
    RwaJarLike(addr).void();
} else if(componentName == URN_COMPONENT) {
    RwaUrnLike(addr).wipe(bal);
} else {
    revert InvalidComponent(ilk, componentName);
}
```

Reply:
The logic being the same is pretty coincidental. There might be jobs in the future that won't have that check, or might have different conditions.

Splitting them like this ensures we keep the same pattern and are not tempted to add more branching to the `work` function.

However I'm not that strong about it. If you and @dbkbali think that's the best approach, I wouldn't oppose.

Reply:
Just as a general point. Smart contract development is not the same as more traditional enterprise code. The focus should almost always be on simple, readable code. This is why we do things like avoid inheritance and copy + paste common snippets such as `auth` inline into each contract. If the number of lines of code can be reduced that should be done.

Reply:
> Smart contract development is not the same as more traditional enterprise code.

Yes, I agree. This was one of the first things I realized when I started writing smart contracts for Ethereum more than 6 years ago.

> The focus should almost always be on simple, readable code.

Again, couldn't agree more. The issue is that different people have different definitions of "simple, readable code". Some people argue that the lesser the number of methods a contract has the better, others believe that contract methods should be concise and do only one thing, even if that increases the amount of code.

> This is why we do things like avoid inheritance [...]

Even in traditional enterprise software inheritance is frowned upon. It's been almost 30 years since people openly [advocate against it](https://en.wikipedia.org/wiki/Composition_over_inheritance). Still, this concept (and other controversial practices) seems to have risen from the ashes with Solidity and its ecosystem. I'm glad to stay away from it :)